### ISSUE TITLE
Provide blocking variants for all queue types

### ISSUE THREAD

**Issue by ChristianWulf** [JCTools/JCTools#40]
I need an SpSc queue that blocks on consumer side if no data is available. Is there a more efficient approach than integrating a semaphore?
For example, the consumer does not always need to check a lock on each consuming action if it caches the queue's size.

**Comment by alexaverbuch** [comment]
+1 on a blocking consumer side
in my case: MpSc, Bounded, Blocking Consumer
the reason is to burn less CPU while consuming events

**Comment by georges-gomes** [comment]
Hi
I submitted a draft pull request in that sense.
My approach is using park/unpark in the most lightweight way I have measured so far.
Let's collaboratively build the best solution!
Cheers
Georges

**Comment by nitsanw** [comment]
This is currently merged into the master branch but under experimental while I experiment with it more closely.

**Comment by ChristianWulf** [comment]
So, we have blocking queues for SPSC and SPMC, but what about MPSC? What could be an efficient solution for this scenario?

**Comment by nitsanw** [comment]
First off, I apologize for delaying this contribution/feature in experimental for so long after the good work delivered by @georges-gomes 

The reason I've been sitting on this  is that I'm unhappy with 2 aspects:
1. The notification cost effects each offer. This is due to the offer being responsible for unparking a blocked consumer.
2. The use of ReentrantLock introduces allocation for waiters on the lock.
The solution I see is to introduce a shared unparking thread who's roll it is to monitor any number of conditions (for the consumer this would be "relaxedPeek() != null" for instance) and trigger an unpark when condition is met. This will need to be considered/measured in the context of many blocked threads waiting on same conditions and so on but should help resolve both issues (and no doubt introduce any number of new issues).
If someone wants to experiment with this approach it would be valuable.

**Comment by ChristianWulf** [comment]
The first point is clear to me and the most important w.r.t. to performance. But what is the problem of a little bit more memory consumption due to a ReentrantLock declaration?

**Comment by nitsanw** [comment]
@ChristianWulf Consumption is fine, churn is an issue
I'm fine with the 'unparker' having some predefined backoff. Make it configurable, but I'm not keen on the producer shouldering the unpark cost. If a consumer has spun down to a park, it's accepted a fairly chunky delay is possible as the worst case.

**Comment by ChristianWulf** [comment]
Do I understand you correctly? You do not want to park producers, but only consumers, right?

**Comment by nitsanw** [comment]
@ChristianWulf this applies on both sides. I don't want to change offer/poll to accomodate a potentially parked counterpart. Empty queues are more common than full queues, so I tend to think of this as a blocked consumer issue, but equally applies on the other side.

**Comment by georges-gomes** [comment]
Hi @ChristianWulf @nitsanw 

TLNR;
My favorite topic ;)
I would be quite keen in promoting the blocking queues variations from "experimental" in order to have a first version of blocking queues [more] available. I ready to finish the missing pieces if there is interest.

----
On our code base at least, a lot of consumers of ArrayBlockingQueue are single threaded.
Simply because we use them as "pipes" and I guess a lot of people are using them this way.
The park strategy used for SC blocking queues have been optimized for "low" latency between the put() and the take(). As far as my testing goes, they are much faster than ArrayBlockingQueue in SC and only slightly faster on MC cases. Benchmarks don't show this because notify() and unpark() are much faster when you run them back to back very quickly. I don't know the details but there is some sort of throughput optimisation here behind the scene.

This being said, there are a few things missing to make it a proper drop-in replacement to ArrayBlockingQueue. (cf. no park strategy for producers on full queues - only offer() + yield() loop for the moment).

-----

@nitsanw I think your idea with a share thread is pretty interesting - I will try to implement it.
But it would not be as reactive between put() and take() unless we agree to "burn" a thread for that. Could be acceptable in some cases but not as a drop-in replacement to Blocking Queues. Nevertheless, to unpark producers and not alter take/poll() performance, I think it's perfect. We don't need so much latency performance when the queue is full. It's nearly a small "back-off" to producer(s) as well...

My 2 cents
GG

**Comment by guidomedina** [comment]
`ChunkedArrayQueue` blocking version will be awesome to have, for my new client to use.

**Comment by akarnokd** [comment]
I'd be somewhat interested in a lower overhead signal/await [around a queue](https://github.com/reactor/reactive-streams-commons/blob/master/src/main/java/rsc/publisher/BlockingIterable.java#L160), however, our use case for blocking is when one really has to leave the reactive world of RxJava/Reactor. By staying, most stages in the processing pipeline is lock-free.

I wonder if using an `AtomicInteger` counter on the offer/poll side could avoid locking/signalling/awaiting that many times (when only the consumer has to block):

```java
Queue<T> queue;
AtomicInteger wip;

public void offer(T v) {
   queue.offer(v);
   if (wip.getAndIncrement() == 0) {
       lock(); signal(); unlock();
   }
}

public T blockingPoll() {
    T v = queue.poll();
    if (v == null) {
       lock(); loopAwait(); unlock();
       v = queue.poll();
    }
    wip.decrementAndGet();
    return v;
}
```

**Comment by SoerenHenning** [comment]
Is there any update on this? I saw there is a `BlockingQueueFactory` in the experimental project. How usable is this already?

**Comment by nitsanw** [comment]
@SoerenHenning this is still experimental... Maybe we can release this as a separate jar with a pre-1.0 version to give people something to work with.

**Comment by guidomedina** [comment]
It would be fun to add something like `queue.poll(50, TimeUnit.MILLIS`) such case is more common than you imagine, I know there are ways to spin and idle but for some projects that's over killing.

The reason for timeout waiting for an element in the queue is to give the blocking queue method an opportunity to check for a condition.

**Comment by nitsanw** [comment]
@guidomedina fun is good :-), I'm currently not sure when I'll have time for this work. So let me outline it if someone else wants to have all the fun:
 - Setup a new module: jctools-blocking
 - Move code from experimental to new module and verify build is good
 - Review current support and document in a README
 - Review testing for any potential holes
 - Release v0.1-beta as a tag (jitpack only release)
 - Release to maven

**Comment by paxel** [comment]
A few insights I learned from using the experimental Blocking template for MpscArrayQueue.

I just took the code template that is meant to do a just in time compiled version and made it manually. With the yieldPut and yieldTake strategie for our proprietary actor system we didn't get the performance that we seeked; probably because the busy loop caused too much noise on the Scheduler (I don't know if this is plausible). With the ParkTake strategy we noticed that the scaling was really bad. We talk about hundred thousands of messages per second loadbalanced to dozends of Actors. As most queues were empty very often, the producers were in Unpark all the time, which seems too expensive. I added a new strategy for take and put that are the same as yield, but call LockSupport.parkNanos(50) instead yield. And that behaves WAAAY better than the other strategies. I don't know what happens to the experimental blocking code. But I'd appreciate if this could go to release at some point. Any thoughts, comments or concerns?

**Comment by guidomedina** [comment]
Maybe take a look at https://github.com/real-logic/agrona/blob/master/agrona/src/main/java/org/agrona/concurrent/IdleStrategy.java
These can really help in cases things like busy spin, yield, etc; maybe you can come up with something good by mixing `IdleStrategy` with blocking queues.

**Comment by franz1981** [comment]
@paxel I believe that a mixed busy, yield and sleep strategy, that would fallback into full park one can guarantee the right balance between throughput/latency and CPU cycles. Please beware pure parkNanos strategies, because they often relies on specific OS settings ie timerslack_ns
and tends to trade better throughputs for worst 99%ile latencies...

**Comment by paxel** [comment]
To be clear here, the actors are never idle. I just turned off the firehose
to see how the Parknano busy wait looks in htop. Then I saw that all actors
were at 3 percent busy take.

I think the park / unpark strategy doesn't scale if you have some (10 to
20) producers that have to send messages to many (40+) very fast consumers
(hardware has plenty of cpus, no problem here). With a high probability,
that the producer has to unpark the consumer (because the queue was empty
before adding) we see actually a decrease in overall throuput with
increasing consumers (jprofiler shows 89 percent unpark).

Changing to the nanopark strategy shows an expected increase in
performance, when increasing consumers.

If nanopark is hardware dependent, then we are lucky, I guess.

**Comment by franz1981** [comment]
If you that many producers probably you are good target to use the new xadd
queue :) I'm pitching you that q eheh)

Il dom 21 lug 2019, 21:04 paxel <notifications@github.com> ha scritto:

**Comment by paxel** [comment]
I am very interested in that xadd. We still lose 3 Gbit/s with the nanopark strategy. Where can I find it?

**Comment by franz1981** [comment]
@paxel here: https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscUnboundedXaddArrayQueue.java

I'm quite sure it will improve a lot after @nitsanw will put his magic touch on it :)
Depending on it, maybe we will have it for the next release, but you can already try it and give your impressions

**Comment by paxel** [comment]
Veeeery interesting. Only one problem: The unbound.

We need limited queues  because of the amount of data. if we have a hickup,
the memory runs full in the blink of an eye if we don't limit the data in
the system.

I can compare the performance, if you are interested, but we can't use it
for production.

Francesco Nigro <notifications@github.com> schrieb am So., 21. Juli 2019,
21:35:

**Comment by franz1981** [comment]
That would great thanks!
Re the unbounded I 100% agree, but considering that this q under burst of many producers can handle several times better latencies/throughput then the ones that use CAS, I would try to sample the size of the q and back-pressure producers in a different way...less precise but it could worth it if your performance bottleneck is the contention between producers..

**Comment by paxel** [comment]
I am terribly sorry. last year went completely under water. I will try to do this asap

**Comment by paxel** [comment]
Ok, I am struggling a bit to use the Xadd queue, as it is only in experimental and no current release on maven central.

But I made comparisons of Put and Take Strategies around MpscArrayQueues.
Proprietary Actor System. 10 entry actors talk to 50 second level actors, that send to 4 actors and those send to 32 actors. This roughly represents a map/reduce architecture.
I generate 1,000,000,000 messages in 10 to 100 threads and round robin them on the first ten actors:

The default YieldPut ScTake Strategy manages 33,000 messages per second
YieldPut YieldTake Strategy manages 2,000,000 messages per second
ParkPut ParkTake Strategy that basically replace yield with parkNanos(50) manages up to 9,000,000 messages per second.

I am not sure if I can test XADD usefully, as I require Blocking Queues and I guess the strategies will hide the actual queue performances.

**Comment by franz1981** [comment]
Hi @paxel 

Now the xadd q Is being released as part of the third release of JCTools: the complex part of using it with the blocking template on JCTools is that that queue isn't chosen by default while using the queue config mechanism with the blocking queue factory :(
It requires some modification of the factory or to implement by hand those...
Will come with a branch you can try that makes it possible to use and it should deliver significant improvements under contention if the load is high enough that the consumer threads won't get asleep.

**Comment by paxel** [comment]
I just found it :D I am stupid. I will test it immediately.

I manually created a Blocking wrapper and just switch out with the MpscUnboundedXaddArrayQueue

**Comment by paxel** [comment]
Ok, I made a half assed attempt to have a limited queue instead of unbound and use an atomic integer.

Even with this shitty attempt, the performance is amazing. The xadd managed 14 to 16 million messages per second, compared to the 9 million messages per second with the MpscArrayQueue. 

with a real limited version that performance would be even better I assume.

**Comment by franz1981** [comment]
@paxel now worries :)
As I have said, it really depends if most of the time is spent while unblocking the consumer thread or while contending on the q offer: only in such cases the xadd q deliver its promises, so maybe the test should be configured/tuned to make that the bottleneck...
But I am happy for the feedback anyway, thanks to have tried it!!

**Comment by franz1981** [comment]
@paxel I've run some synthetic benchmarks on Netty while replacing some queues with the new xadd and getting these results:
![image](https://user-images.githubusercontent.com/13125299/74594576-66b8d000-5038-11ea-803a-c37a5ac40572.png)
```
clq: ConcurrentLinkedQueue
xadd: MpscUnboundedXaddArrayQueue
vyukov: MpscLinkedQueue
cas: MpscUnboundedArrayQueue (default on Netty)
```

**Comment by nitsanw** [comment]
Not happening
